### PR TITLE
Security section [WIP]

### PR DESCRIPTION
**Pull request by RyanSquared**
Requesting review from @aagat

### CODE REVIEW

File: 2-Security/1-intro.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+out onto the internet.
+
+Software nowadays goes through a very long process to make it from the entity
+who made the software, to your computer. In this process, it is often possible
+for someone to slip up and make a mistake, leading to the entire system being
+marked as invalid.
+
+There are certain security measures that you should understand in order to make
+sure that your data is as secure as possible. Throughout this chapter, we'll
+take a look at possible ways to attack, and prevent attacks against, a source
```
Comment (benharri) [inline 2-Security/1-intro.md]: commas unnecessary

File: 2-Security/1-intro.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+minutes? How do you know that, when you go to Facebook, your ISP (internet
+service provider) is not hijacking that domain? Even if you can look at a file
+on one computer and copy it over to another computer, how do you know that your
+flash drive didn't strip out a security measure?
+
+---
+
+In other chapters, we explain things in a format listing "the hard way" and
+"the easy way". We will not be doing that in this chapter. Security is a very
+difficult field to get right, and off-putting for many people.
```
Comment (benharri) [inline 2-Security/1-intro.md]: add "can be" to off-putting

File: 2-Security/1-intro.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
+## Security
+
+Now that you (probably) know some stuff about the internet, it's valuable to
+learn how you can protect yourself from people who want to attack you and your
+services.
+
+A long time ago, people used protocols across local networks to access files.
+Assuming no one could intercept traffic, there was no reason to worry about
+whether or not someone could replace the files with a malicious verson.
```
Comment (Aagat) [inline 2-Security/1-intro.md]:
```suggestion
whether or not someone could replace the files with a malicious version.
```

File: 2-Security/1-intro.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+Obviously, this is no longer the case, and more and more cases continue to rise
+up about how people end up getting hacked and your personal data ends up thrown
+out onto the internet.
+
+Software nowadays goes through a very long process to make it from the entity
```
Comment (Aagat) [inline 2-Security/1-intro.md:19]: This one reads a bit funny.

Maybe?
```suggestion
Software nowadays goes through a very long process to from the entity who made the software to your computer before you can use it.
```

File: 2-Security/1-intro.md (hunk 2 above)
Comment (Aagat) [inline 2-Security/1-intro.md]:
```suggestion
difficult field to get right, and can be off-putting for many people.
```

File: 2-Security/2-hashing.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+copied it from one computer to the other? How do you know that when you take a
+file on the internet and download it, that it's really the file you expect?
+
+Perhaps you've gone to a website and seen a "SHA-1" or "MD5" value. These are
+called a "hash", and are used for checking data integrity. They are what are
+known as a "trapdoor" function: easy to get into, but hard to break out of.
+
+A hash value is, quite simply, a reproducible value generated based off of a
+certain input. It is very hard to find two hash collisions in something that is
+government-grade, such as the SHA-1, SHA-2, and SHA-3 sets of algorithms. It
```
Comment (Aagat) [inline 2-Security/2-hashing.md]: do you really want to say government-grade here? Something like `commonly used` or `widely used` might be better in this context.

File: 2-Security/2-hashing.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+For all these examples, we will be using the bytes `['t', 'e', 's', 't']` and
+as many builtin libraries as possible.
+
+#### Python
+
+```py
+# If using a native string, use .decode() to turn into bytes; hashlib will only
+# accept bytes-like objects.
+import hashlib
+hash = hashlib.sha256(b"test").hexdigest()
```
Comment (Aagat) [inline 2-Security/2-hashing.md:52]: perhaps print the output values as well?

File: 2-Security/2-hashing.md (hunk 5 above)
Comment (RyanSquared) [reply 2-Security/2-hashing.md]: The problem with that is that unfortunately MD5 is still commonly used. I think tying it in with a government standard was a reasonable choice.

File: 2-Security/1-intro.md (hunk 4 above)
Comment (benharri) [reply 2-Security/1-intro.md:19]: remove first "to"

Comment (RyanSquared) [reply 2-Security/1-intro.md:19]: Your suggested fix doesn't read right. I reworded it but "to from the entity" doesn't read right at all.

Comment (Aagat) [reply 2-Security/1-intro.md:19]: The rewording makes it much nicer. :+1:

**Review by benharri — DISMISSED** [review]

**Comment by RyanSquared** [comment]
I did a full AES implementation - from password to padding to encrypting and decrypting.

Requesting review from @lrvick @KellerFuchs to make sure I'm not doing this totally wrong.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+## Table of Contents
+
+### [1-Internet](./1-Internet/)
+#### [1-Internet/1-intro](./1-Internet/1-intro.md)
+#### [1-Internet/2-email](./1-Internet/2-email.md)
+#### [1-Internet/3-irc](./1-Internet/3-irc.md)
+### [2-Security](./2-Security/)
+#### [2-Security/1-intro](./2-Security/1-intro.md)
+#### [2-Security/2-hashing](./2-Security/2-hashing.md)
```
Comment (jjn2009) [inline README.md:14]: this is missing a section

File: 2-Security/3-encryption-symmetric.md [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+import binascii
+import hashlib
+key = hashlib.pbkdf2_hmac('sha256', b'your_password', b'your_salt', 1000000)
+print(binascii.hexlify(key).decode('ascii'))
+```
+
+#### Lua
+
+Unfortunately, as of 2018-10-19, there is no PKCS#5 implementation for luaossl
+as far as the author of this book can tell.
```
Comment (RyanSquared) [inline 2-Security/3-encryption-symmetric.md:67]: s/book/section/

File: README.md (hunk 7 above)
Comment (RyanSquared) [reply README.md:14]: note to self to re-run `generate-readme.py`

File: 2-Security/3-encryption-symmetric.md [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+```py
+import binascii
+import hashlib
+key = hashlib.pbkdf2_hmac('sha256', b'your_password', b'your_salt', 1000000)
+print(binascii.hexlify(key).decode('ascii'))
+```
+
+#### Lua
+
+Unfortunately, as of 2018-10-19, there is no PKCS#5 implementation for luaossl
```
Comment (daurnimator) [inline 2-Security/3-encryption-symmetric.md:66]: Merged today :)

**Comment by RyanSquared** [comment]
Closing this. I'm going to make a PR off of practical usable advice instead.

**Pull request closed without merging**